vortex-array/src/array/chunked/compute/filter.rs

```diff
@@ -0,0 +1,240 @@
+            }
+        }
+    }
+
+    // Now, apply the chunk filter to every slice.
+    for (chunk, chunk_filter) in array.chunks().zip(chunk_filters.iter()) {
+        match chunk_filter {
+            // All => preserve the entire chunk unfiltered.
+            ChunkFilter::All => result.push(chunk),
+            // None => whole chunk is filtered out, skip
+            ChunkFilter::None => {}
+            // Slices => turn the slices into a boolean buffer.
+            ChunkFilter::Slices(slices) => {
+                result.push(filter(&chunk, &slices_to_predicate(slices, chunk.len()))?);
+            }
+        }
+    }
+
+    Ok(result)
+}
+
+/// Filter the chunks using indices.
+fn filter_indices<'a>(
+    array: &'a ChunkedArray,
+    set_indices: Box<dyn Iterator<Item = usize> + 'a>,
+) -> VortexResult<Vec<Array>> {
+    let mut result = Vec::new();
+    let mut current_chunk_id = 0;
+    let mut chunk_indices = Vec::new();
+
+    // Avoid find_chunk_idx and use our own to avoid the overhead.
+    // The array should only be some thousands of values in the general case.
+    let chunk_ends = array.chunk_offsets().into_canonical()?.into_primitive()?;
+    let chunk_ends = chunk_ends.maybe_null_slice::<u64>();
+
+    for set_index in set_indices {
+        let (chunk_id, index) = find_chunk_idx(set_index, chunk_ends);
+        if chunk_id != current_chunk_id {
+            // Push the chunk we've accumulated.
+            if !chunk_indices.is_empty() {
+                let chunk = array
+                    .chunk(current_chunk_id)
+                    .vortex_expect("find_chunk_idx must return valid chunk ID");
+                let filtered_chunk = take(
+                    &chunk,
+                    &PrimitiveArray::from(chunk_indices.clone()).into_array(),
+                )?;
+                result.push(filtered_chunk);
+            }
+
+            // Advance the chunk forward, reset the chunk indices buffer.
+            current_chunk_id = chunk_id;
+            chunk_indices.clear();
+        }
+
+        chunk_indices.push(index as u64);
+    }
+
+    if !chunk_indices.is_empty() {
+        let chunk = array
+            .chunk(current_chunk_id)
+            .vortex_expect("find_chunk_idx must return valid chunk ID");
+        let filtered_chunk = take(
+            &chunk,
+            &PrimitiveArray::from(chunk_indices.clone()).into_array(),
+        )?;
+        result.push(filtered_chunk);
+    }
+
+    Ok(result)
+}
+
+// Mirrors the find_chunk_idx method on ChunkedArray, but avoids all of the overhead
+// from scalars, dtypes, and metadata cloning.
+fn find_chunk_idx(idx: usize, chunk_ends: &[u64]) -> (usize, usize) {
+    let chunk_id = chunk_ends
+        .binary_search(&(idx as u64))
+        .map(|i| if i == chunk_ends.len() { i - 1 } else { i })
+        .unwrap_or_else(|end| end - 1);
+    let chunk_offset = idx - (chunk_ends[chunk_id] as usize);
+
+    (chunk_id, chunk_offset)
+}
```

Comment:
our search_sorted method is defined on primitive arrays and last I checked it wasn't any different (could be different now since 1.81 changed binary_search)

Reply:
It's that the chunk_ends are usually compressed, so decompressing them for every call to find_chunk_idx adds up, along with the scalar/dtype overhead.

Reply:
I meant that you can do this https://github.com/spiraldb/vortex/blob/develop/vortex-array/src/compute/search_sorted.rs#L276-L279

Reply:
i.e. you can do `chunk_ends.search_sorted(&idx, SearchSortedSide::Right)`

Reply:
https://github.com/spiraldb/vortex/blob/develop/vortex-array/src/array/chunked/mod.rs#L98-L101 is the right thing to do

Reply:
Ah yea, that is probably better than what I am doing. Lemme change